src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1755,206 +1906,299 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
 
         public static int Min(this IEnumerable<int> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            Error.CheckSourceNotNull(source);
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Min(this IEnumerable<int?> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            Error.CheckSourceNotNull(source);
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x < value)
-                    value = x;
+                if (!e.MoveNext()) return null;
+                int? cur = e.Current;
+                // Start off knowing that we've a non-null value (or exit here, knowing we don't)
+                // so we don't have to keep testing for nullity.
+                while (!cur.HasValue)
+                {
+                    if (!e.MoveNext()) return cur;
+                    cur = e.Current;
+                }
+                // Keep hold of the wrapped value, and do comparisons on that, rather than
+                // using the lifted operation each time.
+                int valueVal = (value = cur).Value;
+                while (e.MoveNext())
+                {
+                    cur = e.Current;
+                    if (cur.HasValue)
+                    {
+                        int x = cur.Value;
+                        if (x < valueVal)
+                        {
+                            valueVal = x;
+                            value = cur;
+                        }
+                    }
+                }
```

Comment:
Just a minor note here, but if we know that the nullable `HasValue` then we can get a [performance gain](http://ericlippert.com/2012/12/20/nullable-micro-optimizations-part-one/) by changing `int?.Value` to `int?.GetValueOrDefault()`.